### DISCUSSION TITLE
Display an RGB image on top of an image cube

### DISCUSSION THREAD

**Discussion by giswqs** [pyvista/pyvista#6243] [Q&A] [answered]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
import pvxarray
import hypercoast
import numpy as np
import pyvista as pv
import matplotlib.pyplot as plt

url = "https://github.com/opengeos/datasets/releases/download/hypercoast/NEON_D02_SERC_DP3_368000_4306000_reflectance.h5"
filepath = "data/neon.h5"
hypercoast.download_file(url, filepath)

dataset = hypercoast.read_neon(filepath) # Load the dataset as an xarray dataset

da = dataset["reflectance"]  # xarray data array
values = da.to_numpy()
print(values.shape)  # (1000, 1000, 426)

# Select three bands for RGB
band_red = 100  # Replace with the index of your red band
band_green = 60  # Replace with the index of your green band
band_blue = 30  # Replace with the index of your blue band

# Extract the selected bands and replace NaNs with zero
red = np.nan_to_num(values[:, :, band_red])
green = np.nan_to_num(values[:, :, band_green])
blue = np.nan_to_num(values[:, :, band_blue])

# Normalize the bands to [0, 1] for display
def normalize_band(band):
    band_min, band_max = np.min(band), np.max(band)
    return (band - band_min) / (band_max - band_min)

red_normalized = normalize_band(red)
green_normalized = normalize_band(green)
blue_normalized = normalize_band(blue)

# Stack the bands to create an RGB image
rgb_image = np.dstack((red_normalized, green_normalized, blue_normalized))

# Display the RGB image using Matplotlib
plt.imshow(rgb_image)
plt.axis('off')
plt.show()

# Create the spatial reference for the image cube
grid = pv.ImageData()

# Set the grid dimensions: shape because we want to inject our values on the POINT data
grid.dimensions = values.shape

# Edit the spatial reference
grid.origin = (0, 0, 0)  # The bottom left corner of the data set
grid.spacing = (1, 1, 1)  # These are the cell sizes along each axis

# Add the data values to the cell data
grid.point_data["values"] = np.nan_to_num(values).flatten(order="F")  # Flatten the array

# Plot the image cube with the RGB image overlay
p = pv.Plotter()
p.add_mesh(grid, cmap="jet", show_edges=False, clim=(0, 0.5))
# p.add_mesh(rgb_image, show_edges=False, clim=(0, 0.5), show_scalar_bar=False, )

# # Create a texture from the RGB image
texture = pv.numpy_to_texture(rgb_image)

# Overlay the RGB image as a texture on a plane that covers the entire image cube
plane = pv.Plane(center=(values.shape[1] / 2, values.shape[0] / 2, 0), i_size=values.shape[1], j_size=values.shape[0], direction=(0, 0, 1))
p.add_mesh(plane, texture=texture, show_edges=False)

# Show the plot
p.show()
```

### Description

I am trying to render a hyperspectal image with 426 bands using pyvista. In addtion, I want to select three bands from the data and display them as an RGB on top of the image cube. See the complete script below. The problem is that the RGB image is not displayed properly. It is shown as some black patches at the bottom of the image cube. How to display it as an RGB image at the top of the image cube? 

Any advice? @banesullivan 

![image](https://github.com/pyvista/pyvista/assets/5016453/aea8187f-f93d-47d5-b632-b91b8255b8fd)

![image](https://github.com/pyvista/pyvista/assets/5016453/a6952865-ffa3-4cf8-ac4d-47b702445973)

https://github.com/pyvista/pyvista/assets/5016453/43d9e52c-42d6-476d-b3c2-2befe6acc61f

### System Information

```shell
Date: Tue Jun 11 13:04:27 2024 EDT

                  OS : Linux
              CPU(s) : 8
             Machine : x86_64
        Architecture : 64bit
                 RAM : 31.1 GiB
         Environment : Jupyter
         File system : ext4
          GPU Vendor : NVIDIA Corporation
        GPU Renderer : NVIDIA GeForce GTX 1660 SUPER/PCIe/SSE2
         GPU Version : 4.5.0 NVIDIA 550.90.07
    MathText Support : True

  Python 3.10.14 | packaged by conda-forge | (main, Mar 20 2024, 12:45:18)
  [GCC 12.3.0]

             pyvista : 0.43.9
                 vtk : 9.3.0
               numpy : 1.26.4
          matplotlib : 3.8.4
              scooby : 0.10.0
               pooch : 1.8.2
...
       trame_vuetify : 2.5.0
jupyter_server_proxy : 4.1.2
        nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

**Comment by giswqs** [comment]
I tried using the following line to add the RGB image, but it is shown at the bottom of the image cube and only utilizes one band with the default `viridis` colormap. How to display the RGB image at the top of the cube?

```python
p.add_mesh(rgb_image, show_edges=False, clim=(0, 0.5), show_scalar_bar=False)
```
https://github.com/pyvista/pyvista/assets/5016453/92216272-f244-4eae-9cfa-008a4eff6cc0

**Comment by user27182** [comment]
> ```python
> # # Create a texture from the RGB image
> texture = pv.numpy_to_texture(rgb_image)
> 
> # Overlay the RGB image as a texture on a plane that covers the entire image cube
> plane = pv.Plane(center=(values.shape[1] / 2, values.shape[0] / 2, 0), i_size=values.shape[1], j_size=values.shape[0], direction=(0, 0, 1))
> p.add_mesh(plane, texture=texture, show_edges=False)
> ```

It is not necessary to convert the array to a texture and then apply it to a plane. You can simply create a new `ImageData` for the RGB image. To make it 2D set the z dimensions to 1, e.g. :
``` python
x_dim, y_dim = rgb_image.shape[0], rgb_image.shape[1]
z_dim = 1
im = pv.ImageData(dimensions=(x_dim, y_dim, z_dim))

# Add scalar data, you may also need to flatten this
im.point_data["rgb_image"] = rgb_image
```

To move the RGB image to be on top of the cube, set the z-component of its origin to the z_max bound of the cube
``` python
grid_z_max = grid.bounds[5]
im.origin = (0, 0, grid_z_max)
```

EDIT: Fixed the dimensions of the rgb ImageData to match the data array

**Reply by giswqs** [reply]
@user27182 Thank you for your promot response. The rgb_image has a shape of `(1000, 1000, 3)`. The sample code you provided only utilizes the first band, not three bands. It throws the following error:
> ValueError: data length of (1000) != required length (1000000)

**Reply by user27182** [reply]
You need to flatten the array.  Please see the comment I provided in the code from my initial response:
```  python
# Add scalar data, you may also need to flatten this
```

This was something you already did with `grid` in the code you provided:
``` python
# Add the data values to the cell data
grid.point_data["values"] = np.nan_to_num(values).flatten(order="F")  # Flatten the array
```

**Reply by giswqs** [reply]
I did flatten the array, but the length is `1000x1000x3`. Setting `z_dim = 1` results in a length mismatch. 

Setting `z_dim=3` only utilizes the first band, not an RGB image. 
```python
x_dim, y_dim = rgb_image.shape[0], rgb_image.shape[1]
z_dim = 3
im = pv.ImageData(dimensions=(x_dim, y_dim, z_dim))

# Add scalar data, you may also need to flatten this
im.point_data["rgb_image"] = rgb_image.flatten(order="F")

grid_z_max = grid.bounds[5]
im.origin = (0, 0, grid_z_max)
p.add_mesh(im, show_edges=False, show_scalar_bar=False)
```
![image](https://github.com/pyvista/pyvista/assets/5016453/a0e61a77-123b-4257-853a-ea7b0a7e13f6)

**Reply by user27182** [reply, marked as answer]
> I did flatten the array, but the length is `1000x1000x3`. Setting `z_dim = 1` results in a length mismatch.
> 
> Setting `z_dim=3` only utilizes the first band, not an RGB image.

I see. Thanks for clarifying. The confusion may stem from the differences between dimensions and shape. The dimensions refer to physical dimensions in XYZ space. Here we want a 2D image, so the dimensions of `ImageData`  should most certainly be `(1000, 1000, 1)`.

The way the RGB data gets added is with a multi-component scalar array. In this case the array should be two-dimensional with shape `(1 000 000, 3)`. So, I guess what you need here is reshape, not flatten.

Also, you should add `rgb=True` when adding rgb meshes, otherwise the `norm` of the 3-component data will be computed and the default color map will be applied:

``` python
p.add_mesh(im, show_edges=False, show_scalar_bar=False, rgb=True)
```

**Reply by giswqs** [reply]
@user27182 Prefect! It works now. Thank you very much for your help. Really appreciated. 

```python
x_dim, y_dim = rgb_image.shape[0], rgb_image.shape[1]
z_dim = 1
im = pv.ImageData(dimensions=(x_dim, y_dim, z_dim))

# Add scalar data, you may also need to flatten this
im.point_data["rgb_image"] = rgb_image.reshape(-1, 3)*2

grid_z_max = grid.bounds[5]
im.origin = (0, 0, grid_z_max)
p.add_mesh(im, show_edges=False, show_scalar_bar=False, clim=(0, 0.2), rgb=True)
```
![image](https://github.com/pyvista/pyvista/assets/5016453/474b6e1b-611c-4401-8df6-a570feaf3595)

**Reply by user27182** [reply]
@giswqs You're welcome. If you plot rgb images a lot you can also set the array name to `"_rgb"` or `"_rgba"`. This will automatically set `rgb=True` when calling `add_mesh`, which may be helpful.

**Reply by giswqs** [reply]
Great tips! I will certainly use it. Thanks.

**Reply by user27182** [reply]
@giswqs you may also need to add `order='F'` to `reshape`.  If you compare the 1-channel flattened image with `rgb_image.flatten(order="F")`
<img width="521" alt="Screenshot 2024-06-11 at 1 57 35 PM" src="https://github.com/pyvista/pyvista/assets/89109579/4c4d8fb6-0008-496c-b99b-1e742ad788ce">

To the 3-channel image with `rgb_image.reshape(-1, 3)`
<img width="502" alt="Screenshot 2024-06-11 at 1 57 25 PM" src="https://github.com/pyvista/pyvista/assets/89109579/97b78fb2-9cdd-4527-b9ff-01b49abc9778">

You can see that the image is flipped. If you're comparing the image on top to the values from the grid below for some kind of analysis this will give you completely different results. Normally this error is obviously wrong, and the image won't look right (see https://github.com/pyvista/pyvista/issues/4862), but in this case with a symmetric shape this kind of error is not obvious.

**Reply by giswqs** [reply]
@user27182 Thank you very much for pointing this out. I did not notice the flip earlier. I have corrected it. 

https://github.com/opengeos/HyperCoast/pull/49
https://github.com/opengeos/HyperCoast/pull/47
![](https://i.imgur.com/bQqIrQo.png)
